src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -546,40 +546,48 @@ void ShenandoahOldHeuristics::clear_triggers() {
+    // Our intent is to pack old-gen memory into the highest-numbered regions of the heap.  Count all memory
+    // above first_old_region as the "span" of old generation.
+    size_t old_region_span = (first_old_region <= last_old_region)? (num_regions - first_old_region): 0;
+    // Given that memory at the bottom of the heap is reserved to represent humongous objects, the number of
+    // regions that old_gen is "allowed" to consume is less than the total heap size.  The restriction on allowed
+    // span is not strictly enforced.  This is a heuristic designed to reduce the likelihood that a humongous
+    // allocation request will require a STW full GC.
     size_t allowed_old_gen_span = num_regions - (ShenandoahGenerationalHumongousReserve * num_regions) / 100;
 
-    // Tolerate lower density if total span is small.  Here's the implementation:
-    //   if old_gen spans more than 100% and density < 75%, trigger old-defrag
-    //   else if old_gen spans more than 87.5% and density < 62.5%, trigger old-defrag
-    //   else if old_gen spans more than 75% and density < 50%, trigger old-defrag
-    //   else if old_gen spans more than 62.5% and density < 37.5%, trigger old-defrag
-    //   else if old_gen spans more than 50% and density < 25%, trigger old-defrag
-    //
-    // A previous implementation was more aggressive in triggering, resulting in degraded throughput when
-    // humongous allocation was not required.
-
-    size_t old_available = _old_gen->available();
-    size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
-    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_bytes;
+    size_t old_available = _old_gen->available() / HeapWordSize;
+    size_t region_size_words = ShenandoahHeapRegion::region_size_words();
+    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_words;
     assert(old_available >= old_unaffiliated_available, "sanity");
     size_t old_fragmented_available = old_available - old_unaffiliated_available;
 
-    size_t old_bytes_consumed = old_region_count * region_size_bytes - old_fragmented_available;
-    size_t old_bytes_spanned = old_region_span * region_size_bytes;
-    double old_density = ((double) old_bytes_consumed) / old_bytes_spanned;
-
-    uint eighths = 8;
-    for (uint i = 0; i < 5; i++) {
-      size_t span_threshold = eighths * allowed_old_gen_span / 8;
-      double density_threshold = (eighths - 2) / 8.0;
-      if ((old_region_span >= span_threshold) && (old_density < density_threshold)) {
-        trigger_old_is_fragmented(old_density, first_old_region, last_old_region);
-        return;
-      }
-      eighths--;
+    size_t old_words_consumed = old_region_count * region_size_words - old_fragmented_available;
+    size_t old_words_spanned = old_region_span * region_size_words;
+    double old_density = ((double) old_words_consumed) / old_words_spanned;
```

Comment:
... and if `old_region_span = 0`, then you end up with line 573 causing a divide-by-zero error.

Reply:
Good catch.  Thanks.  I'm restructuring the code to only perform this division if old_span_percent > 0.5.  We only compute old_span_percent if ShenandoahGeneratiionalHumongousReserve > 0.

Reply:
Yes, that should fix the issue. Just ping me for a re-review when that is completed. Thanks!